### PR TITLE
Add support for LZ4 compression to reduce CPU consumption during metric collection

### PR DESCRIPTION
**Pull request by mattpatl**
Add support for LZ4 compression as an alternative to zlib compression.

Benchmarks (included) suggest LZ4 could be substantially more CPU-efficient, in exchange for a small penalty on compression ratio. Memory allocations are also reduced. Actual results will depend on the data being sent.

The default compression settings remain as zlib compression level 9, which optimises for compression ratio. Consumers can experiment  by configuring alternative compression settings such as:
```
compression-type=lz4
compression-level=0
```

Benchmark results:
```
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_0-10         	   14469	     74034 ns/op	  735978 B/op	      18 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_1-10         	    9793	    121501 ns/op	 1207731 B/op	      25 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_2-10         	   14373	     77371 ns/op	  814407 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_3-10         	   15165	     75937 ns/op	  814088 B/op	      22 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_4-10         	   14654	     84301 ns/op	  814360 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_5-10         	   16970	     72325 ns/op	  814360 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_6-10         	   17054	     71677 ns/op	  814360 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_7-10         	   15502	     71669 ns/op	  814360 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_8-10         	   16933	     75193 ns/op	  814360 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Zlib/zlib_compression_level_9-10         	   16864	     72691 ns/op	  814360 B/op	      23 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_0-10           	  692482	      1723 ns/op	    1374 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_1-10           	   47011	     24341 ns/op	    1245 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_2-10           	   49320	     24484 ns/op	    1301 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_3-10           	   48962	     24378 ns/op	    1431 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_4-10           	   46981	     24449 ns/op	    1535 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_5-10           	   49279	     24729 ns/op	    1194 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_6-10           	   49369	     24387 ns/op	    1109 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_7-10           	   46466	     24398 ns/op	    1314 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_8-10           	   46854	     24713 ns/op	    1312 B/op	       9 allocs/op
BenchmarkHttpForwarderV2Compression_Lz4/lz4_compression_level_9-10           	   49504	     25004 ns/op	    1215 B/op	       9 allocs/op
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/atlassian/gostatsd
 
-go 1.19
+go 1.21.4
```
Comment (mattpatl) [inline go.mod:3]: I have not upgraded the go version. This has been adjusted to match the go version in other files (e.g. Docker, README, and CI config).

File: pkg/statsd/handler_http_forwarder_v2.go [hunk 2]
```diff
@@ -27,11 +26,14 @@ import (
...
 	"github.com/atlassian/gostatsd/pkg/stats"
 	"github.com/atlassian/gostatsd/pkg/transport"
+	"github.com/atlassian/gostatsd/pkg/web"
 )
 
 const (
 	defaultConsolidatorFlushInterval = 1 * time.Second
 	defaultCompress                  = true
+	defaultCompressionType           = "zlib"
+	defaultCompressionLevel          = 9
```
Comment (mattpatl) [inline pkg/statsd/handler_http_forwarder_v2.go:36]: These defaults match the existing behaviour.

File: pkg/statsd/handler_http_forwarder_v2.go [hunk 3]
```diff
@@ -134,6 +144,14 @@ func NewHttpForwarderHandlerV2(
 	if concurrentMerge <= 0 {
 		return nil, fmt.Errorf("concurrent-merge must be positive")
 	}
+	compressionType, err := web.ReadCompressionType(compressionTypeStr)
+	if err != nil {
+		return nil, fmt.Errorf(err.Error())
```
Comment (MovieStoreGuy) [inline pkg/statsd/handler_http_forwarder_v2.go:149]: You can simplify this by returning `err` here, `fmt.Errorf` is more for additional written context to help understand the issue. For example:

```go
return nil, fmt.Errorf("unable to read compression: %w")
```

File: pkg/web/http_receiver_v2_test.go [hunk 4]
```diff
@@ -24,157 +24,177 @@ import (
 func TestForwardingEndToEndV2(t *testing.T) {
 	t.Parallel()
 
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (mattpatl) [inline pkg/web/http_receiver_v2_test.go:27]: This diff is particularly difficult to read. This large red block has simply been nested inside the `t.Run` call (see larg green block below).

**Review by MovieStoreGuy — APPROVED** [review]

**Pull request merged**